### PR TITLE
bpo-42603: Use pkg-config to get TCL/TK paths for tkinter.

### PR DESCRIPTION
**Pull request by m000**
When paths are not supplied with `--with-tcltk-includes` and `--with-tcltk-libs`, use pkg-config (if available) to get the values of `TCLTK_INCLUDES` and `TCLTK_LIBS`. Makes easier to compile Python with a non-standard Tcl/Tk.

https://bugs.python.org/issue42603

### CODE REVIEW

**Comment by E-Paine** [comment]
Thank you for this patch. Please modify `configure.ac` and then regenerate `configure` using Autoconf (see the [devguide](https://devguide.python.org/setup/#regenerate-configure))

**Comment by m000** [comment]
@E-Paine Done. Autoreconf also makes modifications unrelated to the actual change. I've left those out.

**Comment by E-Paine** [comment]
I'm not really the person to be reviewing this, but maybe @serhiy-storchaka or @ned-deily you would like to take a look?

**Comment by m000** [comment]
If accepted, it would also make sense to backport this patch to a few versions back. Are separate PRs required for this, or will it be handled by a maintainer?

**Comment by E-Paine** [comment]
I personally believe this to be an enhancement which is not typically backported. If the core who merges decides otherwise, they'll add labels and let one of the bots cherry-pick it.

**Comment by m000** [comment]
@E-Paine Can you poke the Azure Pipelines test? IIRC, it succeeded the first time, but failed after only rewording the NEWS.d file.

**Comment by E-Paine** [comment]
@m000 Unfortunately, I cannot do so as I have the same permissions as you. It is an unrelated asyncio failure, though, so can just be ignored.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by m000** [comment]
Is it normal for the bot to mark "awaiting merge" requests as "stale"?

**Comment by ned-deily** [comment]
"stale" status is a recent addition to our workflow meaning a PR that is awaiting review or other action for some time. Sorry for the delay, I do intend to review it in the near future. And thanks for your contribution.

**Review by ned-deily — CHANGES_REQUESTED** [review]
This seems to me to be a generally useful enhancement. However, it does introduce a potentially subtle change in behavior on macOS: if you have Tck and Tk frameworks installed in `/Library/Frameworks` and you have Tcl and Tk installed elsewhere along with a `pkg-config` that references it (like, say, in a MacPorts environment), by default Python will now try to link with the Tcl/Tk pointed to by `pkg-config` rather than the `/Library/Frameworks` version.  The behavior can be overriden on configure (`./configure  --with-tcltk-includes=  --with-tcltk-libs=`) or make (`make TCLTK_INCLUDES= TCLTK_LIBS=`). Since the primary documentation for all this is in the top-level `setup.py`, I think the comments there should be updated to reflect the changes in this PR.

**Comment by terryjreedy** [comment]
Manolis, from past experience, the bogus review requests are likely from force-pushing.  That should not be needed and sometimes messes up a PR even worse.   Just push your branch to your fork.

**Comment by m000** [comment]
I have made the requested changes; please review again.

I have updated the comments in `setup.py` to reflect the changes as accurately as possible. This includes renaming `detect_tkinter_explicitly()` to `detect_tkinter_fromenv(self)`, since the use of `pkg-config` is not explicit.

The subtle behaviour change for building tkinter on macOS can be easily be inferred by the updated comments. However, I didn't make an explicit note for it _in the comments_ but instead updated the NEWS.d entry (seems more appropriate).

For easier review, I put the new changes in a different commit. If needed, I can squash before merging.

Sorry for the bogus review requests!

File: configure.ac
Comment (tiran) [inline configure.ac:3166]: Could you please use autotools macros instead of manually running pkg-config?

https://autotools.io/pkgconfig/pkg_check_modules.html

**Review by tiran — CHANGES_REQUESTED** [review]

File: configure.ac
Comment (m000) [reply configure.ac:3166]: This would be inconsistent with the rest of `configure.ac`. I believe this change should be made in a separate PR for all the uses of pkg-config.

**Comment by terryjreedy** [comment]
Please do not squash.  Reviewers can look at both individual commits, especially when designed for easier review, and the overall change (the 'Files changed' button at the top).  I know I do that, and would here if I were reviewing this.  I even make separate commits, for my own review, when making substantial PRs.  As the big green button below says, we squash just before merging into cpython.

File: configure.ac
Comment (m000) [reply configure.ac:3166]: @tiran I've checked in more detail about using `PKG_CHECK_MODULES` instead of the current approach.

I would be against using `PKG_CHECK_MODULES` for the following reasons:

1. Introduces an inconsistency with the way libffi is configured (also runs pkg-config manually).
2. Introduces a naming inconsistency because of the `TCLTK_CFLAGS` variable that will be created. We already have `TCLTK_INCLUDES` which is essentially the same. Also, openssl uses `OPENSSL_INCLUDES`. `XXX_CFLAGS` is not used for any other prerequisite.
3. Introduces a naming conflict because of the `TCLTK_LIBS` variable. The variable is already defined, which creates confusion.
4. Replaces ~10L of straightforward and arguably robust code (the equivalent libffi code hails from 2010) with >100L of autogenerated clutter.

To address 1-3, changes that go beyond the scope of this PR need to be introduced. If 1-3 are not addressed, the resulting configuration script will include inconsistencies that will have to be addressed before they get worse.

Given this, I believe it is better to go with manually running pkg-config for now. The use of `PKG_CHECK_MODULES` should be discussed in a separate issue, so the changes are applied in a uniform way so that there are no inconsistencies in the produced configure script.

Let me know what you think. Anyone else wishing to weigh-in, please let us know of your thoughts on this.

Comment (ned-deily) [reply configure.ac:3166]: I agree with @m000 here.  Tcl/Tk version selection has various legacy behaviors especially on macOS.  Trying to convert this to use a configure time check with an autotools macros would break compatibility.  I would be more in favor of using such an approach if and when we re-write the current Unix-y build system to address third-party library selection for all modules.

Comment (m000) [reply configure.ac:3166]: No further updates from @tiran. Marking discussion as resolved since @ned-deily agrees with my points above.

**Comment by m000** [comment]
I have made the requested changes; please review again.

- Changes requested by @ned-deily implemented.
- Changes requested by @tiran deferred for resolution in another PR.

**Review by ned-deily — APPROVED** [review]
Looks good to me. Thanks for the updates!

**Pull request merged**
